src/dsp_tools/utils/xmlupload/xmlupload.py

```diff
@@ -242,63 +336,39 @@ def _upload_resources(
+        resources=resources,
+        imgdir=imgdir,
+        preprocessing_done=preprocessing_done,
+    )
+    bitstream_size_uploaded_mb = 0.0
 
     for i, resource in enumerate(resources):
         resource_start = datetime.now()
         filetype = ""
         filesize = round(bitstream_all_sizes_mb[i], 1)
         bitstream_duration_ms = None
+
         resource_iri = resource.iri
         if resource.ark:
             resource_iri = convert_ark_v0_to_resource_iri(resource.ark)
 
-        # in case of a multimedia resource: upload the multimedia file
-        resource_bitstream = None
-        if preprocessing_done and resource.bitstream:
-            resource_bitstream = resource.get_bitstream(resource.bitstream.value, permissions_lookup)
-        elif resource.bitstream:
-            pth = resource.bitstream.value
-            try:
-                bitstream_start = datetime.now()
-                filetype = Path(pth).suffix[1:]
-                img: Optional[dict[Any, Any]] = try_network_action(
-                    sipi_server.upload_bitstream,
-                    filepath=str(Path(imgdir) / Path(pth)),
-                )
-                bitstream_duration = datetime.now() - bitstream_start
-                bitstream_duration_ms = bitstream_duration.seconds * 1000 + int(bitstream_duration.microseconds / 1000)
-                mb_per_sec = round((filesize / bitstream_duration_ms) * 1000, 1)
-                metrics.append(
-                    MetricRecord(resource.id, filetype, filesize, "bitstream upload", bitstream_duration_ms, mb_per_sec)
-                )
-            except BaseError as err:
-                err_msg = err.orig_err_msg_from_api or err.message
-                msg = f"Unable to upload file '{pth}' of resource '{resource.label}' ({resource.id})"
-                print(f"WARNING: {msg}: {err_msg}")
-                logger.warning(msg, exc_info=True)
-                failed_uploads.append(resource.id)
-                continue
-            bitstream_size_uploaded_mb += bitstream_all_sizes_mb[i]
-            msg = f"Uploaded file '{pth}' ({bitstream_size_uploaded_mb:.1f} MB / {bitstream_size_total_mb} MB)"
-            print(msg)
-            logger.info(msg)
-            internal_file_name_bitstream = img["uploadedFiles"][0]["internalFilename"]  # type: ignore[index]
-            resource_bitstream = resource.get_bitstream(internal_file_name_bitstream, permissions_lookup)
+        upload_bitstream_successful, resource_bitstream = _check_for_bitstream_and_upload_if_there(
+            resource=resource,
+            sipi_server=sipi_server,
+            imgdir=imgdir,
+            filesize=filesize,
+            bitstream_size_uploaded_mb=bitstream_size_uploaded_mb,
+            permissions_lookup=permissions_lookup,
+            metrics=metrics,
+            bitstream_size_total_mb=bitstream_size_total_mb,
+            preprocessing_done=preprocessing_done,
```

Comment:
boolean parameters to a function are generally a red flag: They indicate that the function will have a top level if-else (which is the case here), so the function does two fundamentally different things. Ergo, it should be two separate functions.